rfcs/0096-embedded-rust-interface-api.md

```diff
@@ -0,0 +1,274 @@
+
+This requires an additional build step in `build.rs` to reference the existing C files using `cc-rs`:
+
+```rust
+use cc;
+fn main() {
+    cc::Build::new()
+        .include("../codegen/host/include")
+        .include("../runtime/include")
+        .file("../codegen/host/src/default_lib0.c")
+        .file("../codegen/host/src/default_lib1.c")
+        .compile("mlf");
+}
+```
+
+This builds the existing C code and C interface for reference in Rust using the `ffi` module:
+
+```rust
+
+```
+
+## C Wrapper Structs
+By wrapping the C types in Rust structs, and exposing a `new` constructor it allows us to take a defined Rust array and translate it into a `void*`. Users of this interface only need to deal with pure Rust:
+
+```rust
+/// Input tensors for TVM module "rusty_coffee"
+#[repr(C)]
+pub struct Inputs {
+    input: *mut ::std::os::raw::c_void,
+}
+
+impl Inputs {
+    pub fn new <'a>(
+        input: &mut [u8; 100],
+    ) -> Self {
+        Self {
+            input: input.as_ptr() as *mut ::std::os::raw::c_void,
+        }
+    }
+}
+```
+
+For devices, we can only store the void pointer as it's passed without type information through the C runtime:
+```rust
+/// Device context pointers for TVM module "rusty_coffee"
+#[repr(C)]
+pub struct Devices {
+    ethos_u: *mut ::std::os::raw::c_void,
+}
+
+impl Devices {
+    pub fn new <'a>(
+        ethos_u: *mut ::std::os::raw::c_void,
+    ) -> Self {
+        Self {
+            ethos_u: ethos_u,
+        }
+    }
+}
+```
+
+## C Wrapper Entrypoint
+Similar to the above, using the Rust FFI we can create an entrypoint function which passes the void pointers directly into C FFI from the Rust structs with the appropriate `unsafe` block to prevent user facing code from having to be `unsafe`. The `run` wrapper also checks the return code of TVM and converts it into a [standard Rust `Result` object](https://doc.rust-lang.org/rust-by-example/error/result.html).
+
+```rust
+/// Entrypoint function for TVM module "rusty_coffee"
+/// # Arguments
+/// * `inputs` - Input tensors for the module
+/// * `outputs` - Output tensors for the module
+pub fn run(
+    inputs: &mut Inputs,
+    outputs: &mut Outputs,
+) -> Result<(), ()> {
+    unsafe {
+        let ret = tvmgen_rusty_coffee_run(
+            inputs,
+            outputs,
+        );
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+}
+
+extern "C" {
+    pub fn tvmgen_rusty_coffee_run(
+        inputs: *mut Inputs,
+        outputs: *mut Outputs,
+    ) -> i32;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This introduces a second embedded API alongside the C API, but fundamentally we want users to be able to create applications in the language that best suits their needs.
+
+Whilst it is an innovation project, this will be supported with best efforts but may be lag in features behind the C interface API.
```

Comment:
I don't believe we do this currently for the C Embedded API and if we want to, we should define it as part of the over-arching microTVM project?

Reply:
My question is mainly about scope of support of the current proposal just to bring some clarity, that is all.

Reply:
I don't think we can provide that at the interface level, it is the same as the Embedded C interface which relies on the various functions of the AOT executor and C runtime to function - it's not specific to the Embedded Rust Interface?

Reply:
This particular comment do not have things to do with interface, but instead the models that can be supported and possible restrictions. 

In this case stating that "this particular module will support all the models that can be supported through embedded C API, because there is a gap in feature X, we might miss models X, Y, Z" would be sufficient.